Add requestTrackingAuthorization() API to request user's tracking consent on iOS 14+

BendingBender opened the pull request:
Adds the [bespoke](https://github.com/chemerisuk/cordova-plugin-idfa/pull/5#issuecomment-705427700) `requestTrackingAuthorization()` API.

Fixes #7

BendingBender committed:
Add requestTrackingAuthorization() API to request user's tracking consent on iOS 14+

chemerisuk commented:
@BendingBender thanks for changes. I don't have strong complains, just couple of naming suggestions. We could probably publish a new major release with those changes, so my notes:

1. Rename `requestTrackingAuthorization` to `requestPermission`
2. Rename `trackingTransparencyStatus` to `trackingPermission`
3. Rename `limitAdTracking` to `trackingLimited`
4. Use `undefined` value instead of `"NotAvailable"` for `trackingPermission`

Feel free to split the task into multiple PRs.
In result we could use boilerplate below:

```js
cordova.plugins.idfa.getInfo(info => {
    if (!info.trackingLimited) {
        console.log(info.idfa || info.aaid);
    } else if (info.trackingPermission === "NotDetermined") {
        cordova.plugins.idfa.requestPermission(result => {
            if (result === "Authorized") {
                cordova.plugins.idfa.getInfo(info => {
                    console.log(info.idfa || info.aaid);
                });
            }
        });
    }
});
```

What do you think?

BendingBender commented:
Yeah, I was thinking the same when I was writing the code. Didn't know that you were open for a major version bump.
I'll rewrite the pull request.

BendingBender commented:
Regarding
> Use undefined value instead of "NotAvailable" for trackingPermission

I'm not a big fan of `undefined`/nullish values. I find it much more practical and more readable to have explicit values if something is not available, ideally of the same type as the real values. And I find it very annoying that most cordova plugins expose platform API differences even if it's easy to consolidate them into a single API for all platforms. This is why my first PR contained this kind of consolidation code.

I know this is just my preference but I would rather introduce the `NotAvailable` state to both the iOS and Android interface instead of replacing the `NotAvailable` value with undefined. What do you think?

BendingBender committed:
Consolidate API

- limitAdTracking -> isTrackingLimited
- trackingTransparencyStatus -> trackingPermission
- requestTrackingAuthorization -> requestPermission
- remove `NotAvailable` permission status, replace it with `null`
- add TrackingPermission enum for easier permission value lookup

BendingBender commented:
I used `null` instead of `undefined` beacause it was easier to pass `null` directly from native code. Otherwise I'd have to convert `null` values to `undefined` on JS side.

chemerisuk commented:
@BendingBender if you want to define constants for `trackingPermission` values then I'd suggest to send `ATTrackingManagerAuthorizationStatus` as an integer numbers without number-to-readable-string conversion. Should simplify iOS implementation:

```js
- (void)requestPermission:(CDVInvokedUrlCommand *)command {
    [self.commandDelegate runInBackground:^{
        if (@available(iOS 14, *)) {
            [ATTrackingManager requestTrackingAuthorizationWithCompletionHandler:^(ATTrackingManagerAuthorizationStatus status) {
                CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsInt:status];
                [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
            }];
        } else {
            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }
    }];
}
```

On js side values for constant values will be numbers instead of strings.

chemerisuk commented on README.md:
```diff
@@ -22,7 +22,9 @@
-Use variable `ANDROID_PLAY_ADID_VERSION` to override dependency version on Android.
+Use variable `ANDROID_PLAY_ADID_VERSION` to override dependency version on Android:
+
+    $ cordova plugin add cordova-plugin-idfa --variable ANDROID_PLAY_ADID_VERSION='16.+'
```
Good comment. I should probably add such example into each repo because people are asking a lot about this.

chemerisuk commented on README.md:
```diff
@@ -32,21 +34,73 @@ The API is available on the `cordova.plugins.idfa` global object.
+            return info.idfa || info.aaid;
+        } else if (info.trackingPermission === TrackingPermission.NotDetermined) {
+            return idfaPlugin.requestPermission().then(result => {
+                if (result !== TrackingPermission.Authorized) {
```
Can be simplified:
```js
if (result === TrackingPermission.Authorized) {
    return idfaPlugin.getInfo().then(info => {
                    return info.idfa || info.aaid;
                });
}
```

BendingBender replied:
Well, this was on purpose. This way we short-circuit invalid state handling and leave the valid state handling outside the if-block. Still want me to change this?

chemerisuk replied:
> leave the valid state handling outside the if-block

I think valid state handling should be in the latest `then` block (line 101).

BendingBender replied:
I don't really understand what you mean. We can't move the check in line 91 to line 102, the call to `idfaPlugin.getInfo()` depends on this check. There is still a notion of validity there, it is basically invalid/unnecessary to call `idfaPlugin.getInfo()` if `result !== idfaPlugin.TRACKING_PERMISSION_AUTHORIZED`.

Could you please elaborate on this?

chemerisuk replied:
Sorry I misunderstood. I was thinking not about valid handler, but success handler:
```js
cordova.plugins.idfa.getInfo(info => {
    if (!info.trackingLimited) {
        return info.idfa || info.aaid;
    } else if (info.trackingPermission === "NotDetermined") {
        return cordova.plugins.idfa.requestPermission(result => {
            if (result === TrackingPermission.Authorized) {
                return cordova.plugins.idfa.getInfo(info => {
                    return info.idfa || info.aaid;
                });
            }
        });
    }
}).then(idfaOrAaid => {
    if (idfaOrAaid) {
        // success handler
    }
});
```

So we don't really need "to leave the valid state handling outside the if-block." because the success handler for both cases is later anyway.

chemerisuk commented on src/android/IdfaPlugin.java:
```diff
@@ -21,7 +21,7 @@ protected void getInfo(CallbackContext callbackContext) throws Exception {
         JSONObject result = new JSONObject();
         result.put("aaid", info.getId());
-        result.put("limitAdTracking", info.isLimitAdTrackingEnabled());
+        result.put("isTrackingLimited", info.isLimitAdTrackingEnabled());
```
I'd still prefer `trackingLimited` name because it's a field, `isTrackingLimited` looks like a method name.

chemerisuk commented on www/Idfa.js:
```diff
@@ -2,9 +2,22 @@ var exec = require("cordova/exec");
 var PLUGIN_NAME = "Idfa";
 
 module.exports = {
+    TrackingPermission: Object.freeze({
```
In JS word more conventional to have all constant uppercased and without introducing `enum` objects:

```js
module.exports = {
    TRACKING_PERMISSION_AUTHORIZED: 1,
    TRACKING_PERMISSION_DENIED: 2,
    ...
    getInfo: function() {
    ...
}
```

BendingBender commented:
@chemerisuk Any opinions on this: https://github.com/chemerisuk/cordova-plugin-idfa/pull/9#issuecomment-754747501?

BendingBender committed:
Changes after review

- rename `isTrackingLimited` -> `trackingLimited`
- return tracking permission status as numerical value
- use more DOM-like interface to define constants for the numerical status

chemerisuk commented:
@BendingBender my thoughts were to avoid having `trackingPermission` key for env where it's not supported (currently iOS < 14):

```objectivec
- (void)getInfo:(CDVInvokedUrlCommand *)command {
    [self.commandDelegate runInBackground:^{
        NSString *idfaString = [[[ASIdentifierManager sharedManager] advertisingIdentifier] UUIDString];
        BOOL enabled = [[ASIdentifierManager sharedManager] isAdvertisingTrackingEnabled];
        CDVPluginResult* pluginResult;

        if (@available(iOS 14, *)) {
            // workaround, as long as Apple deferred the roll-out of manadatory tracking permission popup
            // we'll assume that if the idfa string is not nullish, then it's allowed by user
            if (!enabled && idfaString != nil && ![@"00000000-0000-0000-0000-000000000000" isEqualToString:idfaString]) {
                enabled = YES;
            }

            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:@{
              @"idfa": idfaString,
              @"trackingLimited": [NSNumber numberWithBool:!enabled],
              @"trackingPermission": @(ATTrackingManager.trackingAuthorizationStatus)
            }];
        } else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:@{
              @"idfa": idfaString,
              @"trackingLimited": [NSNumber numberWithBool:!enabled]
            }];
        }
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}
```

If there is no key we don't need to discuss which value to choose `null` or something else :) Android implementation doesn't have this key as well so it's consistent too.

BendingBender commented:
> If there is no key we don't need to discuss which value to choose null or something else :) Android implementation doesn't have this key as well so it's consistent too.

My question was related to the fact that we return `null`/`undefined` at all. We could instead introduce a real value for the "not available" status, `"NotAvailable"` in the previous implementation, `-1` in the current. And we would return this value on all platforms, no need to check whether it's iOS or Android. This simplifies a lot of handling and allows to have a single implementation for both platforms on the user side. I would also implement a handler for `requestPermission()` on Android which always returns the "not available" status, this way we would have a unified interface on all platforms.

I know this is not how most cordova plugins are but as a cordova user it annoys me every time I have to paper over plaform differences exposed by plugins although it would be easy for the plugins to provide a unified interface.

The only difference that really makes sense in my opinion is that the `idfa` and `aaid` properties are separated, these are actually different things and they need to stay separated.

I was also thinking of introducing a convenience layer method that basically implements the code in the example. The implementations is kind of unwieldy and I expect most users to not need to make separate calls to `requestPermission()` and to `getInfo()`. So we should expose kind of a `getInfoWithPermissionDialog()` (or something like that) helper that makes it easier to use this plugin.

chemerisuk commented:
> I have to paper over platform differences exposed by plugins although it would be easy for the plugins to provide a unified interface.

For some APIs providing unified interfaces can be a bad thing. Because when you polyfill a native API that even does not have an equivalent (like `trackingPermission` does not have any sense for Android platform or iOS < 14) it could be real mess for developers. Therefore I'd rather have `trackingPermission` key only for platforms where it makes sense and drop it everywhere else.

By default unimplemented methods in cordova are rejected with an error. So we don't need to do anything on Android.

I know the pain with feature check to detect if a method implemented or not. For this plugin

```js
if (info.trackingPermission === idfaPlugin.TRACKING_PERMISSION_NOT_DETERMINED)
```
also does the feature check for method `requestPermission()`, because on unsupported platforms `trackingPermission` is always `undefined`.

> I was also thinking of introducing a convenience layer method that basically implements the code in the example

Let's discuss it in a separate issue probably.

chemerisuk commented on src/ios/IdfaPlugin.m:
```diff
@@ -40,11 +22,26 @@ - (void)getInfo:(CDVInvokedUrlCommand *)command {
+            }];
+        } else {
+            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
+            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
```
Lets send an error response here with message "requestPermission is supported only for iOS >= 14"

BendingBender committed:
Changes after review

- change example code
- return an error when `requestPermission` is called on iOS <14

BendingBender commented:
> Therefore I'd rather have trackingPermission key only for platforms where it makes sense and drop it everywhere else.
>
> By default unimplemented methods in cordova are rejected with an error. So we don't need to do anything on Android.

OK, if you find the example code appealing, then I give up.

chemerisuk commented:
@BendingBender great work, thanks!

BendingBender commented:
When do you intend to release a new version?

chemerisuk commented:
@BendingBender new version 2 is available now.
